reasoner/planner/AnswerCountEstimator.java

```diff
@@ -0,0 +1,486 @@
+        return estimators.get(conjunction).estimateAnswers(variableFilter, includedResolvables);
+    }
+
+    private void registerConjunctionAndInitialize(ResolvableConjunction conjunction) {
+        if (!estimators.containsKey(conjunction)) {
+            estimators.put(conjunction, new ConjunctionAnswerCountEstimator(this, conjunction));
+        }
+
+        if (estimators.get(conjunction).mayInitialize()) {
+            initializationStack.add(conjunction);
+            estimators.get(conjunction).initialize();
+            assert initializationStack.get(initializationStack.size() - 1) == conjunction;
+            initializationStack.remove(initializationStack.size() - 1);
+
+            if (cycleHeads.size() == 1 && cycleHeads.contains(conjunction)) {
+                // Reset all cyclic paths
+                for (ResolvableConjunction conjunctionToReset : toReset) {
+                    estimators.get(conjunctionToReset).reset();
+                }
+            }
+        }
+    }
+
+    private void reportInitializationCycle(ResolvableConjunction conjunction) {
+        cycleHeads.add(conjunction);
+        for (int i = initializationStack.size() - 1; initializationStack.get(i) != conjunction; i--) {
+            toReset.add(initializationStack.get(i));
+        }
+    }
+
+    private static class ConjunctionAnswerCountEstimator {
+
+        private final ResolvableConjunction conjunction;
+        private final AnswerCountEstimator answerCountEstimator;
+        private final LogicManager logicMgr;
+        private final AnswerCountModel answerCountModel;
+        private Map<Variable, LocalEstimate> unaryEstimateCover;
+        private Map<Resolvable<?>, List<LocalEstimate>> retrievableEstimates;
+        private Map<Resolvable<?>, LocalEstimate> inferrableEstimates;
+        private Map<Resolvable<?>, LocalEstimate> unaryEstimates;
+
+        private long fullAnswerCount;
+        private long negatedsCost;
+
+        private enum InitializationStatus {NOT_STARTED, RESET, IN_PROGRESS, COMPLETE}
+
+        private InitializationStatus initializationStatus;
+
+        public ConjunctionAnswerCountEstimator(AnswerCountEstimator answerCountEstimator, ResolvableConjunction conjunction) {
+            this.answerCountEstimator = answerCountEstimator;
+            this.answerCountModel = answerCountEstimator.answerCountModel;
+            this.logicMgr = answerCountEstimator.logicMgr;
+            this.conjunction = conjunction;
+            this.fullAnswerCount = -1;
+            this.negatedsCost = -1;
+            this.initializationStatus = InitializationStatus.NOT_STARTED;
+        }
+
+        private FunctionalIterator<Resolvable<?>> resolvables() {
+            return iterate(logicMgr.compile(conjunction));
+        }
+
+        private boolean isInferrable(Resolvable<?> resolvable) {
+            return resolvable.isConcludable() &&
+                    !logicMgr.applicableRules(resolvable.asConcludable()).isEmpty();
```

Comment:
in the context of where you're using this, you're always getting the resolvables from the `compile()` method. This means we've already filtered out the concludables that do not have applicable rules! This means this method can simply be replaced by `resolvable.isConcludable()`